server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java

```diff
@@ -283,11 +341,155 @@ private void listStartedShards(
+            logger.debug("flushing {} primary fetching requests", queuedRequests.size());
+            final CountDownLatch nodeRequestLatch = new CountDownLatch(queuedRequests.size());
+            for (DiscoveryNode node : queuedRequests.keySet()) {
+                Map<ShardId, ShardRequestInfo<NodeGatewayStartedShards>> shardRequests = queuedRequests.get(node);
+                Map<ShardId, String> shards = new HashMap<>();
+                for (ShardId shardId : shardRequests.keySet()) {
+                    shards.put(shardId, shardRequests.get(shardId).getCustomDataPath());
+                }
+
+                assert shards.isEmpty() == false;
+                logger.debug("Batch sending number of {} primary shard async fetch requests to node {}", shards.size(), node.getId());
+
+                // send shards fetch request per node
+                final var curNodeRequests = queuedRequests.get(node);
+                client.executeLocally(
+                    TransportNodesBatchListGatewayStartedShards.TYPE,
+                    new TransportNodesBatchListGatewayStartedShards.Request(shards, new DiscoveryNode[] { node }),
+                    new ActionListener<>() {
+                        @Override
+                        public void onResponse(NodesGatewayBatchStartedShards nodesBatchStartedShards) {
+                            nodeRequestLatch.countDown();
+                            if (nodesBatchStartedShards.failures().size() > 0) {
+                                // single node, got failed node request then node response must be empty.
+                                assert nodesBatchStartedShards.getNodes().size() == 0;
+                                for (var request : curNodeRequests.entrySet()) {
+                                    request.getValue()
+                                        .getListener()
+                                        .onResponse(
+                                            new NodesGatewayStartedShards(
+                                                nodesBatchStartedShards.getClusterName(),
+                                                new ArrayList<>(), // empty response
+                                                nodesBatchStartedShards.failures()
+                                            )
+                                        );
+                                }
+                                return;
+                            }
+
+                            assert nodesBatchStartedShards.getNodes().size() == 1;
+                            NodeGatewayBatchStartedShards startedShardResponse = nodesBatchStartedShards.getNodes().get(0);
+                            List<NodeGatewayBatchStartedShard> startedShards = startedShardResponse.getStartedShards();
+                            assert startedShards.size() == curNodeRequests.size();
+                            for (NodeGatewayBatchStartedShard shard : startedShards) {
```

Comment:
@howardhuanghua  What will happen if the node which the response from has shard throws  ElasticSearch Exception and didn't add to batchStartedShards at code `batchStartedShards.addStartedShard(listStartedShard(entry.getKey(), entry.getValue()));` in `TransportNodesBatchListGatewayStartedShards.java`.  
Can ignore them directly  ?

Reply:
Add fix to check the total target shards number.